### PR TITLE
Tidy up codebase

### PR DESCRIPTION
**Pull request by gregtyler**
Split up code into clearer functions, bulk out localstack initialization, add tests

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -32,85 +34,139 @@ type MyResponse struct {
...
-func HandleLambdaEvent(event ObjectCreatedEvent) (MyResponse, error) {
-  bucketName := event.Records[0].S3.Bucket.Name
-  objectKey := event.Records[0].S3.Object.Key
+type LambdaTagValues struct {
+	pass string
+	fail string
+}
 
-	sess := session.Must(session.NewSession())
+type DownloaderInterface interface {
```
Comment (hawx) [inline main.go:42]: `Downloader` instead

File: main.go [hunk 2]
```diff
@@ -32,85 +34,139 @@ type MyResponse struct {
...
 	if err != nil {
 		log.Printf("failed to write tags, %v", err)
-		return MyResponse{}, fmt.Errorf("failed to write tags, %v", err)
+		return fmt.Errorf("failed to write tags, %v", err)
+	}
+
+	return nil
+}
+
+func (l Lambda) HandleEvent(event ObjectCreatedEvent) (MyResponse, error) {
```
Comment (hawx) [inline main.go:129]: I'd probably put all these methods on `*Lamdba` instead, then use `&Lambda{}` to create

File: main.go [hunk 3]
```diff
@@ -32,85 +34,139 @@ type MyResponse struct {
...
+func (l Lambda) tagFile(bucket string, key string, status string) error {
+	tagging, err := l.s3.GetObjectTagging(&s3.GetObjectTaggingInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(key),
 	})
 
 	if err != nil {
 		log.Printf("failed to get tags, %v", err)
-		return MyResponse{}, fmt.Errorf("failed to get tags, %v", err)
+		return fmt.Errorf("failed to get tags, %v", err)
```
Comment (hawx) [inline main.go:107]: you can use `%w` for errors, then they can be unwrapped

Comment (hawx) [reply main.go:107]: and you could move the logging up to `HandleEvent` from these methods so it isn't duplicated as much

File: main.go [hunk 4]
```diff
@@ -32,85 +34,139 @@ type MyResponse struct {
...
+		return fmt.Errorf("failed to download file, %v", err)
 	}
 
-	// scan with clamav
-	status := "ok"
-	cmd := exec.Command("./bin/clamscan", filename)
+	return nil
+}
+
+func (l Lambda) scanFile() (string, error) {
```
Comment (hawx) [inline main.go:79]: might make more sense to have this not on Lambda, as `func scanFile(path string) (bool, error)` (or create a `Scanner` type if you want to mock it)

File: main_test.go [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+	"io"
+	"testing"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/stretchr/testify/assert"
+)
+
+type mockDownloader struct {
```
Comment (hawx) [inline main_test.go:14]: I know we hand-rolled all this at the start, but maybe we should move to using `github.com/stretchr/testify/mock` now. It's being used on search-service and we already use the `assert` package. Although I like that you don't need to learn another package to figure out what is going on, the manualness can be a bit tiring.

File: scanner.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+			} else {
+				return false, fmt.Errorf("failed to scan file, %w", err)
+			}
+		} else {
+			return false, fmt.Errorf("failed to scan file, %w", err)
+		}
+	}
+
+	return pass, nil
+}
```
Comment (hawx) [inline scanner.go:27]: I think you could probably simplify it to this

```go
if err := cmd.Run(); err != nil {
  if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() == 1 {
    return false, nil
  } 

  return false, fmt.Errorf("failed to scan file, %w", err)
}

return true, nil
```

File: test.sh [hunk 7]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env bash
+set -euo pipefail
+
+docker compose up --wait localstack
+docker-compose exec -T localstack bash -c '. /scripts/wait/wait-until-s3-ready.sh'
+docker-compose exec -T localstack bash -c 'echo "Test file" | awslocal s3 cp - s3://uploads-bucket/valid.txt'
+docker-compose exec -T localstack bash -c 'echo "X5O!P%@AP[4\PZX54(P^)7CC)7}\$EICAR-STANDARD-ANTIVIRUS-TEST-FILE!\$H+H*" | awslocal s3 cp - s3://uploads-bucket/invalid.txt'
```
Comment (hawx) [inline test.sh:7]: cool, I was hoping there would be a nice test value

File: main.go [hunk 8]
```diff
@@ -2,115 +2,156 @@ package main
...
+	bucketName := event.Records[0].S3.Bucket.Name
+	objectKey := event.Records[0].S3.Object.Key
+
+	err := l.downloadFile(bucketName, objectKey)
+	if err != nil {
+		log.Print(err)
+		return MyResponse{}, err
+	}
+
+	// scan with clamav
```
Comment (hawx) [inline main.go:115]: these comments are kind of pointless when I can read the method names

File: main_test.go (hunk 5 above)
Comment (gregtyler) [reply main_test.go:15]: Done, turns out that's a really nice library.

File: test.sh (hunk 7 above)
Comment (gregtyler) [reply test.sh:7]: I can't wait for Veracode to false-positive us for having a virus in the repo 😅

**Review by hawx — APPROVED** [review]

**Pull request merged**